### PR TITLE
Improve config type

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Change:

```python
    config: Optional[Dict[str, Any]] = None
```

to:

```python
    config: Optional[
        Union[
            OpenAIConfig,
            AnthropicConfig,
            DatabricksConfig,
            MLflowConfig,
            CustomConfig,
        ]
    ] = None
``` 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/config.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from pathlib import Path
 from pydantic import BaseModel, validator, parse_obj_as, Extra, ValidationError
 from pydantic.json import pydantic_encoder
-from typing import Optional, Union, List, Dict, Any
+from typing import Optional, Union, List
```
Comment (harupy) [inline mlflow/gateway/config.py:9]: /ai Explain this change.

Comment (harupy) [inline mlflow/gateway/config.py:9]: /ai Explain this change.

File: mlflow/gateway/providers/openai.py [hunk 2]
```diff
@@ -5,22 +5,25 @@
...
-        config = self.config.model.config
-        token = config["openai_api_key"]
-        headers = {"Authorization": f"Bearer {token}"}
-        if org := config.get("openai_organization"):
+        headers = {"Authorization": f"Bearer {self.openai_config.openai_api_key}"}
+        if org := self.openai_config.openai_organization:
             headers["OpenAI-Organization"] = org
         async with aiohttp.ClientSession(headers=headers) as session:
-            url = "/".join([config["openai_api_base"].rstrip("/"), path.lstrip("/")])
+            url = "/".join([self.openai_config.openai_api_base.rstrip("/"), path.lstrip("/")])
```
Comment (harupy) [inline mlflow/gateway/providers/openai.py:26]: /ai Summarize this change.

File: mlflow/gateway/config.py [hunk 3]
```diff
@@ -147,16 +155,13 @@ def validate_provider(cls, value):
...
+        if provider := values.get("provider"):
             config_type = config_types[provider]
-            config_instance = config_type(**config)
+            return config_type(**config)
 
-            return config_instance
-        else:
-            raise MlflowException.invalid_parameter_value(
-                "A provider must be provided for each gateway route."
-            )
```
Comment (harupy) [inline mlflow/gateway/config.py:159]: /ai What's the main change in this diff?

File: mlflow/gateway/config.py [hunk 4]
```diff
@@ -147,16 +155,13 @@ def validate_provider(cls, value):
...
+        raise MlflowException.invalid_parameter_value(
+            "A provider must be provided for each gateway route."
+        )
```
Comment (harupy) [inline mlflow/gateway/config.py:164]: /ai summarize this change

File: mlflow/gateway/config.py [hunk 5]
```diff
@@ -137,7 +137,15 @@ def _resolve_api_key_from_input(api_key_input):
...
-    config: Optional[Dict[str, Any]] = None
+    config: Optional[
+        Union[
+            OpenAIConfig,
+            AnthropicConfig,
+            DatabricksConfig,
+            MLflowConfig,
+            CustomConfig,
+        ]
+    ] = None
```
Comment (harupy) [inline mlflow/gateway/config.py:148]: /ai how does pydantic handle `Union` type?

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Pull request merged**